### PR TITLE
LexicalMarkdownPlugin: say hello to markdown exports

### PR DESCRIPTION
**Pull request by mansimransingh**
This is a first pass at implementing Markdown support for lexical. 

It depends on Apple's swift-markdown package. 

It needs tests and covering more edges cases (embedded lists, indentations, etc) - but is enough to get started with.

I am also working on a markdown import for our use case, which I will add to this plugin in the next day or so.

### CODE REVIEW

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  // Also, no support for checkmarks in Lexical AFAIK.
+
+  public func exportMarkdown() throws -> Markdown.BlockMarkup {
+    let children = getChildren().exportAsBlockMarkdown()
+      .compactMap { $0 as? Markdown.ListItem }
+    switch getListType() {
+    case .bullet:
+      return Markdown.UnorderedList(children)
+    case .check:
+      // TODO (mani) - how does lexical mark a checked item?
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:89]: it doesn't yet! Checklists are still on the todo list.

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    return node
+  }
+}
+
+extension LexicalListPlugin.ListNode: NodeMarkdownBlockSupport {
+  // NOTE (mani) - there are some oddities when converting lists
+  // especially when lists have sub lists.
+  // Sometimes Lexical will not realise that the top level list has been deleted
+  // and so it will look like `List -> ListItem -> List -> [ListItem]` which outputs
+  // incorrect markdown. Assume indentations are not properly supported.
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:79]: Indentations _should_ be correctly supported, but you may have discovered a bug. I'll check with the web guys whether they've made any fixes to the list logic which might need porting to Swift.

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    Markdown.Link(destination: getURL(),
+                  getChildren()
+      .exportAsInlineMarkdown(indentation: getIndent())
+      .compactMap { $0 as? Markdown.RecurringInlineMarkup })
+  }
+}
+
+extension Lexical.CodeNode: NodeMarkdownBlockSupport {
+  public func exportMarkdown() throws -> Markdown.BlockMarkup {
+    // TODO (mani) - do code blocks have formatting?
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:132]: They don't currently have formatting. It's by design that you can't apply e.g. bold/italic to the contents of a code block. Eventually we'd like to add a syntax highlighter to code blocks, similar to what web lexical has.

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+extension Lexical.ParagraphNode: NodeMarkdownBlockSupport {
+  public func exportMarkdown() throws -> Markdown.BlockMarkup {
+    return Markdown.Paragraph(getChildren().exportAsInlineMarkdown(indentation: getIndent()))
+  }
+}
+
+extension Lexical.TextNode: NodeMarkdownInlineSupport {
+  public func exportMarkdown(indentation: Int) throws -> Markdown.InlineMarkup {
+    let format = getFormat()
+    var node: Markdown.InlineMarkup = Markdown.Text(makeIndentation(indentation) + getTextPart())
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:35]: I'm not quite sure what your logic for indentation is doing here, in particular how it interacts with block vs inline markup. (It doesn't help that I'm not familiar with the particular markdown library you're using, I guess!)

My main question is: given indentation is a property of element nodes (and used only when they're `isInline()==false`), how come you're applying the indentation in the inline markup section? (I might be missing something so let me know if so!)

Consider a paragraph node that contains two text nodes: "abc" and "def". Maybe the second one is bold, that's why there's two. The paragraph is indentation level 2.

What we'd want to output is

```
\t\tabc*def*
```

but wouldn't we get

```
\t\tabc\t\t*def*
```

if we use this logic?

File: Plugins/LexicalMarkdown/LexicalMarkdownTests/LexicalMarkdownTests.swift [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  override func tearDown() {
+    lexicalView = nil
+  }
+
+  func testNodesToMarkdown() throws {
+    guard let editor else { XCTFail(); return }
+
+    XCTFail("Unimplemented")
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdownTests/LexicalMarkdownTests.swift:32]: Can we throw in an `XCTExpectFailure` until this test is is implemented? (I won't be able to land something with a failing test, as our internal CI will object.)

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  String(repeating: "\u{009}", count: count)
+}
+
+public protocol NodeMarkdownBlockSupport: Lexical.Node {
+  func exportMarkdown() throws -> Markdown.BlockMarkup
+}
+
+public protocol NodeMarkdownInlineSupport: Lexical.Node {
+  func exportMarkdown(indentation: Int) throws -> Markdown.InlineMarkup
+}
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:24]: I'm wondering whether the names of the functions should be `exportBlockMarkdown()` and `exportInlineMarkdown()`, just to make it super clear without having to cross reference to the protocol names?

File: Plugins/LexicalMarkdown/LexicalMarkdown/LexicalMarkdown.swift [hunk 7]
```diff
@@ -0,0 +1,32 @@
+//
+//  LexicalMarkdown.swift
+//
+//
+//  Created by mani on 15/11/2023.
+//
```
Comment (amyworrall) [inline Plugins/LexicalMarkdown/LexicalMarkdown/LexicalMarkdown.swift:6]: Can you use the MIT license header here? (Our linter will fix it when I try to land, so no biggie.)

**Review by amyworrall — COMMENTED** [review]
Overall I love where this is going. 

Do you happen to know how/if the Swift Markdown library allows customising the precise syntax generated by built in types e.g. Strong or Emphasis? I ask because I have a surface that I could try using this plugin (Workplace Chat composer), but it uses a slightly custom version of Markdown (single underlines => italic, single asterisk => bold).

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift (hunk 2 above)
Comment (mansimransingh) [reply Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:79]: I plan to work on indentations once I have importing working. I'm on a time crunch to get a POC shipped for a project 🤞

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift (hunk 4 above)
Comment (mansimransingh) [reply Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:35]: You're right. This is a bug. Consecutive inline nodes should only have indentation applied once. That should be easily achievable if i treat `LineBreakNode` as a block level node. Will review.

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift (hunk 6 above)
Comment (mansimransingh) [reply Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:24]: Good call! Will update.

**Comment by mansimransingh** [comment]
Swift markdown supports a whole set of formatting options which make it easy to customise the output:

https://apple.github.io/swift-markdown/documentation/markdown/markupformatter/options/

If it doesn't support your exact variant, it should be relatively straight forward to have a fork.

One of my current challenges is having support for underline & italics - which are mutually exclusive (combined under Emphasis node) in swift markdown.

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift (hunk 2 above)
Comment (amyworrall) [reply Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:79]: I know that feeling!

File: Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift (hunk 4 above)
Comment (amyworrall) [reply Plugins/LexicalMarkdown/LexicalMarkdown/NodeMarkdownSupport.swift:35]: FWIW I have never been entirely happy with how LineBreakNode works. It's there for compatibility with web Lexical. But it does make various things more complicated than IMO it needs to be!

**Comment by mansimransingh** [comment]
@amyworrall im yet to fix the indentation issues. Everything else is addressed.

for my project, I went another route. The javascript library has markdown import/export support and we can rely on the fact that JSON internals exported from JS are compatible with swift library (besides this fix https://github.com/facebook/lexical-ios/pull/51). 

So i packed the javascript library along with 2 utility functions (convert markdown -> JSON and JSON -> markdown), and using the Javascript headless lexical editor, over a [JSContext](https://developer.apple.com/documentation/javascriptcore/jscontext) - i can do the format conversions. This turned out to be quicker than implementing the markdown import reliably.

If you're interested, i can create a branch for that too.

**Comment by amyworrall** [comment]
> So i packed the javascript library along with 2 utility functions (convert markdown -> JSON and JSON -> markdown), and using the Javascript headless lexical editor, over a [JSContext](https://developer.apple.com/documentation/javascriptcore/jscontext) - i can do the format conversions. This turned out to be quicker than implementing the markdown import reliably.
> 
> If you're interested, i can create a branch for that too.

That sounds interesting from a tech perspective! For a long term Lexical iOS perspective, we probably do want to do things natively.

Did you make an attempt at the native Markdown import? If so, would you mind sharing what issues you ran into? That may save me some time when I attempt to look at it :)

**Comment by mansimransingh** [comment]
I added a live markdown view in the playground and started work on imports on a branch here:

https://github.com/mansimransingh/lexical-ios/tree/markdown-importing

Just have to go through each `[MarkupVisitor](https://github.com/mansimransingh/lexical-ios/blob/ca2fce9b38cb987713154c0c7371aae27b3c5f6d/Plugins/LexicalMarkdown/LexicalMarkdown/MarkdownImporter.swift#L14)` function and implement correctly / map it to a lexical node.

It stacks on top of this PR.

**Comment by amyworrall** [comment]
> I added a live markdown view in the playground and started work on imports on a branch here:

Awesome, will check it out.

BTW sorry for the churn on the "imported this pull request" messages -- having some trouble wiht our internal tooling.

**Comment by amyworrall** [comment]
hi! Have merged. Had to change the name of the SwiftMarkdown library, we call it SwiftMarkdown internally not just Markdown. Haven't checked that still works with the playground, will check tomorrow.

**Pull request closed without merging**